### DISCUSSION TITLE
using for_each instead of this module?

### DISCUSSION THREAD

**Discussion by doesnotexist** [DacoDev/terraform-module-proxmox-lxc#4] [Q&A] [answered]
Hi,
I tried your module out and it helped me understand Terraform modules. So thank you. 
As I got further along I realized I could just use for_each and locals instead. 
Have you considered doing that? Perhaps I was using this module for a different purpose than you intended.
https://developer.hashicorp.com/terraform/language/meta-arguments/for_each

Here's my example of creating 2 containers using for_each with the out of the box bpg/proxmox provider:

```
terraform {
  required_providers {
    proxmox = {
      source = "bpg/proxmox"
    }
  }
}

locals {
  container_template_file     = "http://download.proxmox.com/images/system/debian-12-standard_12.2-1_amd64.tar.zst"
  container_template_checksum = "761b3b52ba296c3e396b79f89bdb30fed121d2353a641f84d633a87897ed6d19"
  container_distro            = "debian"
  network_bridge              = "vmbr0"
  node_name                   = "proxmox1"
  containers = {
    "test1" = {
      ram_MiB                 = 4096
      swap_MiB                = 512
      vm_id                   = 1000
      cpu_cores               = 2
      container_template_id   = proxmox_virtual_environment_file.container_template.id
    }
    "test2" = {
      ram_MiB                 = 8192
      swap_MiB                = 0
      vm_id                   = 1001
      cpu_cores               = 4
      container_template_id   = proxmox_virtual_environment_file.container_template.id
    }
  }
}

resource "proxmox_virtual_environment_file" "container_template" {
  content_type = "vztmpl"
  datastore_id = "local"
  node_name    = "proxmox1"
  source_file {
    path     = local.container_template_file
    insecure = true
    checksum = local.container_template_checksum
  }
}

resource "proxmox_virtual_environment_container" "proxmox_lxc" {
  description = "Managed by Terraform"
  for_each = local.containers

  node_name   = local.node_name
  vm_id       = each.value.vm_id
  unprivileged  = true

  initialization {
    hostname = each.key
    ip_config {
      ipv4 {
        address = "dhcp"
      }
    }

    user_account {
      keys = [
        trimspace(tls_private_key.lxc_container_key.public_key_openssh)
      ]
      password = random_password.lxc_password.result
    }
  }

  cpu {
    architecture = "amd64"
    cores        = each.value.cpu_cores
  }

  memory {
    dedicated = each.value.ram_MiB
    swap      = each.value.swap_MiB
  }

  disk {
    datastore_id = "local-zfs"
    size         = "10"
  }

  network_interface {
    name   = "veth0"
    bridge = local.network_bridge
  }

  operating_system {
    template_file_id = proxmox_virtual_environment_file.container_template.id
    type             = local.container_distro
  }
}

resource "random_password" "lxc_password" {
  length           = 16
  override_special = "_%@"
  special          = true
}

resource "tls_private_key" "lxc_container_key" {
  algorithm = "RSA"
  rsa_bits  = 2048
}

output "container_password" {
  value     = random_password.lxc_password.result
  sensitive = true
}

output "container_private_key" {
  value     = tls_private_key.lxc_container_key.private_key_pem
  sensitive = true
}

output "container_public_key" {
  value = tls_private_key.lxc_container_key.public_key_openssh
}
```

**Comment by DacoDev** [comment, marked as answer]
Hey, glad I could help!

I haven't revisited this in a long time - when I wrote this initially I probably didn't have a good grasp of counts and for_each at least compared to now.

What you've proposed works fine, however it retains the complexity where the user would need to write and maintain all of those resources themselves. 

In my mind, a module is used to abstract all of the individual components and return to the user a sane-default resource, and I use the variable inputs extensively to validate the inputs that ProxMox is documented as having.

With modules, you can source it from multiple places, in one repo or many, without having to duplicate all the 7 or so resources. If you need to update something in one of the resources, instead of updating it everywhere you've copied the stack of resources, you'd just need to update the module at the source for all the usages to get the update.

So, with your way it works, especially if you need more control over specific resources, but if you had a simple use-case, you could call my module with just the 6 inputs.

**Reply by DacoDev** [reply]
Just ran into a good example of module features helping; in this case the input validation.

I was testing my module since it's been so long and when I went to set `container_template_file` I copy-pasted `alpine-3.18-default_20230607_amd64.tar.xz` instead of the actual link `http://download.proxmox.com/images/system/alpine-3.18-default_20230607_amd64.tar.xz` and TF prevented me from running the plan kicking back the error:

```
│ Error: Invalid value for variable
│ 
│   on proxmox-lxc.tf line 5, in module "proxmox-lxc":
│    5:   container_template_file = "alpine-3.18-default_20230607_amd64.tar.xz"
│     ├────────────────
│     │ var.container_template_file is "alpine-3.18-default_20230607_amd64.tar.xz"
│ 
│ URL should start with HTTP or HTTPS and end with one of .tar.gz, .tar.zst, or .tar.xz.
│ 
│ This was checked by the validation rule at .terraform/modules/proxmox-lxc/variables.tf:21,3-13.
```
